### ISSUE TITLE
Network Support

### ISSUE THREAD

**Issue by LinuxDonald** [fifengine/fifengine#994]
It would be nice to add fife network support for rpg games for example for other types of games.

**Comment by prock-fife** [comment]
I agree but I would want to use an open source library to accomplish this (and make it optional).   We would need to investigate this further.  Any suggestions for an API?

**Comment by vtchill** [comment]
I know raknet is a big one that i've heard of in the past. Not sure of other alternatives or the state of it currently. It may be one to look into.

**Comment by LinuxDonald** [comment]
[Rakenet](https://github.com/OculusVR/RakNet) Dead?
[Enet](https://github.com/lsalzman/enet) Dead? (Python Support)
[Zeroc-ice](https://github.com/zeroc-ice/ice) active (Python Support)
[PCOCO](https://github.com/pocoproject/poco) active

**Comment by prock-fife** [comment]
[Asio](http://think-async.com/Asio)
[Boost.asio](http://www.boost.org/doc/libs/1_63_0/doc/html/boost_asio.html)

**Comment by jakoch** [comment]
> It would be nice to add fife network support for rpg games for example for other types of games.

One line of a networking related feature request equals 1000 hours of complicated work. :)

- Raknet's state isn't good since Oculus acquired it. It's now old and unmaintained. 👎
- Zeroc ICE would cause license issues. https://zeroc.com/licensing LGPL + GPLv2 might fit for open-source, but when fifengine includes zeroc and a user decides to build a commerical game with it, then he would need to buy a commerical license. 👎 
- Asio++ - boost license. haven't seen this in games.
- Boost.asio - boost license. cross-platform. 
- Enet is MIT licensed and used in lots of games ([Red Eclipse](https://github.com/red-eclipse/base/tree/master/src/enet), [0AD](https://github.com/0ad/0ad/tree/2dd6032b7dccf033bbbaef4390e8a250cd69be16/source/network)), works cross-platform and has a Python wrapper ([pyenet](https://pypi.python.org/pypi/pyenet)) Additional note: UH uses pyenet already. 

I would go with Enet to keep it simple.  👍 

server: create_server, ip, port, max_clients, maybe bandwidth control in/out
client: create_client, ip, port, maybe bandwidth control in/out
packet compression and/or serialization modes

Apart from library selection, there are also many ways to implement this.
A standalone server with connecting clients. Or a star-like networking approach, where one gaming client opens the server (hub) for other gaming clients to connect.

How to support multiple gaming types?
The game using fifengine decides, which pieces of information to exchange with a another client via a server. I think Enet combined with [protobuf](https://developers.google.com/protocol-buffers/) would be a good fit. That would give us a flexible approach to describe the format of the exchanged message content. Of course, non structured data (e.g. a StringStream) is easier for debugging packets.

closing connection, Jens

**Comment by prock-fife** [comment]
@jakoch sounds like you have a fair bit of knowledge when it comes to games and network code.  Is ENet supported on andoid/ios?

**Comment by jakoch** [comment]
> Is ENet supported on andoid/ios?

iOS is basically unix-style network stuff and enet should work right out of the box.

For Android, uhm, that's Java land and complicated. The library needs to be wrapped via [Android NDK](https://developer.android.com/ndk/guides/concepts.html). 
That enables your Android game (Java code) to call functions of the native library through the Java Native Interface framework. There is a repo around: https://github.com/julienr/libenet-android

Another way would be to use a native Java networking lib, e.g. Ion or [RetroFit](https://github.com/square/retrofit) (which has Protobuf support) and just re-implement the package handling.

**Comment by LinuxDonald** [comment]
@jakoch will you maybe do this? (add network support to fife)

**Comment by jakoch** [comment]
No. I see this as a feature request with a really low priority and a milestone "far in the future".
Please don't underestimate the time and effort needed to get a networking layer implemented, tested, debugged and profiled. Considering that we have zero test coverage and many longstanding issues to resolve, i won't start to work on new features with this kind of workload.
